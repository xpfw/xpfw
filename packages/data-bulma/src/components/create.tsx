import * as React from "react"

import { dataOptions,  ICreateHookProps, useCreateWithProps } from "@xpfw/data"
import { getMapToFromProps, iterateSubFields, prependPrefix, SharedField } from "@xpfw/form"
import { get } from "lodash"
import { observer } from "mobx-react-lite"

const BulmaCreate: React.FunctionComponent<ICreateHookProps> = observer((props) => {
  const createProps = useCreateWithProps(props)
  const fields: any[] = []
  iterateSubFields(props.schema, (key, schema) => {
    fields.push(<SharedField key={key} schema={schema} prefix={prependPrefix(getMapToFromProps(props), props.prefix)} />)
  })
  const gotErr = createProps.error != null
  const result = createProps.state
  let msg: any
  if (gotErr) {
    msg = (
      <div className="notification is-danger">
        Error please recheck your inputs or connection {JSON.stringify(createProps.error)}
      </div>
    )
  }
  if (result) {
    msg = (
      <div className="notification is-success">
        Successfully created {get(result, dataOptions.idPath)}
      </div>
    )
  }
  return (
    <div>
      {fields}
      <a className="button is-primary" onClick={createProps.submitCreate}>Create</a>
      {msg}
    </div>
  )

})

export default BulmaCreate
export {
  ICreateHookProps
}
