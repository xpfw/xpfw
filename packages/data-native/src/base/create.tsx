import * as React from "react"
import { dataOptions,  useCreateWithProps, ICreateHookProps } from "@xpfw/data"
import { getMapToFromProps, iterateSubFields, prependPrefix, SharedField } from "@xpfw/form"
import { get } from "lodash"
import { observer } from "mobx-react"
import { View } from "react-native"
import { Button, Text, Card } from "react-native-elements"

const NativeCreate: React.FunctionComponent<ICreateHookProps> = observer((props) => {
  const createProps = useCreateWithProps(props)
  const fields: any[] = []
  iterateSubFields(props.schema, (key, schema) => {
    fields.push(<SharedField key={key} schema={schema} prefix={prependPrefix(getMapToFromProps(props), props.prefix)} />)
  })
  const gotErr = createProps.error != null
  const result = createProps.state
  let msg: any
  if (gotErr) {
    msg = (
      <Card
        containerStyle={{backgroundColor: "green", margin: 3}}
      >
        <Text style={{color: "white", textAlign: "center"}}>
          Error please recheck your inputs or connection {JSON.stringify(createProps.error)}
        </Text>
      </Card>
    )
  }
  if (result) {
    msg = (
      <Card
        containerStyle={{backgroundColor: "green", margin: 3}}
      >
        <Text style={{color: "white", textAlign: "center"}}>
          Successfully created {get(result, dataOptions.idPath, "")}
        </Text>
      </Card>
    )
  }
  return (
    <View>
      {fields}
      <View style={{marginTop: 20, marginLeft: 5, marginRight: 5}}>
        <Button onPress={createProps.submitCreate} title="Create" />
      </View>
      {msg}
    </View>
  )
})

export default NativeCreate