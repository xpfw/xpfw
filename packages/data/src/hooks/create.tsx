import { ExtendedJSONSchema, FormStore, getMapTo, jsonValidator, memo, prependPrefix } from "@xpfw/form"
import { get, isNil } from "lodash-es"
import { useEffect } from "react"
import DbStore from "../store/db"
import UserStore from "../store/user"

const submitCreate = (schema: ExtendedJSONSchema, mapTo?: string, prefix?: string) => {
  return async () => {
    if (mapTo == null) {
      mapTo = getMapTo(schema, mapTo)
    }
    const newValue = FormStore.getValue(mapTo, prefix)
    jsonValidator.validate(schema, newValue)
    FormStore.setError(mapTo, jsonValidator.errors, prefix)
    const err = FormStore.getError(mapTo, prefix)
    let res
    if (err != null) {
      res = await DbStore.create(schema, prefix)
    }
    return res
  }
}
const u: any = DbStore
const useCreate = (schema: ExtendedJSONSchema, mapTo?: string, prefix?: string, reset?: boolean) => {
  if (mapTo == null) { mapTo = getMapTo(schema, mapTo) }
  useEffect(() => {
    if (reset) {
      u.createState[prependPrefix(mapTo, prefix)] = null
      FormStore.setValue(mapTo, {}, prefix)
      FormStore.setError(mapTo, undefined, prefix)
    }
  }, [reset])
  return {
    error: FormStore.getError(mapTo, prefix),
    state: DbStore.getCreateState(mapTo, prefix),
    loading: FormStore.getLoading(mapTo, prefix),
    user: UserStore.getUser(),
    submitCreate: memo(() => submitCreate(schema, mapTo, prefix), [JSON.stringify(schema), mapTo, prefix])
  }
}

export default useCreate
export {
  submitCreate
}
