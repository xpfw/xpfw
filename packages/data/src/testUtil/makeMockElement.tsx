import { cloneDeep, get, isString, set } from "lodash-es"
import * as React from "react"

const idtmp: string[] = []

const makeMockElement = (name: string, stripIdsAt?: string | string[], idPath: string = "_id", stringifier?: () => any) => {
  return class extends React.Component<any, any> {
    public render() {
      const toStringify = stringifier ? stringifier() : this.props
      const stripMe = Array.isArray(stripIdsAt) ? stripIdsAt : [stripIdsAt]
      if (Array.isArray(stripMe)) {
        for (const stripper of stripMe) {
          if (isString(stripper)) {
            let toStrip = get(this.props, stripper)
            if (!Array.isArray(toStrip)) {
              toStrip = [toStrip]
            }
            const newIds = []
            for (const obj of toStrip) {
              const id = isString(obj) ? obj : get(obj, idPath)
              let newId = idtmp.indexOf(id)
              if (newId === -1) {
                newId = idtmp.length + 1
                idtmp.push(id)
              }
              const objClone = cloneDeep(obj)
              if (isString(obj)) {
                newIds.push(newId)
              } else {
                set(objClone, idPath, newId)
                newIds.push(objClone)
              }
            }
            if (newIds.length > 0 && newIds[0] != null) {
              set(toStringify, stripper, newIds)
            }
          }
        }
      }
      return (
        <div className={name}>
          {JSON.stringify(toStringify, undefined, 2)}
        </div>
      )
    }
  }
}

export default makeMockElement
