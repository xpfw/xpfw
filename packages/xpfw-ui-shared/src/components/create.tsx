import { FormErrorStore, FormStore, LoadingStore } from "@xpfw/form-shared"
import { filterFields, globals, IField, IForm, IFormError } from "@xpfw/validate"
import { get, isNil } from "lodash"
import * as React from "react"
import { ComponentBase } from "resub"
import DbStore, { IResult } from "../store/db"
import UserStore from "../store/user"

export interface ISharedFormCreate extends React.Props<any> {
  prefix?: string
  method?: string
  /**
   * Will reset the form values
   */
  resetOnMount?: boolean
  /**
   * Will reset the result of the last creation
   */
  resetState?: boolean
  form: IForm
}

export interface ISharedFormCreateState {
  error?: IFormError
  state?: any
  user?: any
  loading: boolean
}

export type SubmitCreate = () => Promise<any>

export interface IFormCreateProps extends ISharedFormCreate, ISharedFormCreateState {
  submitCreate: SubmitCreate
  /** Prefiltered Fields to display */
  fields: IField[]
}

const submitCreate = (thisRef: {props: ISharedFormCreate}) => {
  return async () => {
    const prefix = get(thisRef, "props.prefix", "")
    await FormStore.validateForm(get(thisRef, "props.form"),
      get(thisRef, "props.method"), prefix, get(thisRef, "state.user"))
    const err =
    FormErrorStore.getError(`${prefix && prefix.length > 0 ? prefix + "." : ""}${get(thisRef, "props.form.model")}`)
    let res
    if (!isNil(err) && err.ok) {
      res = await DbStore.create(get(thisRef, "props.form"), prefix)
    }
    return res
  }
}

function SharedFormCreate<T>(Component: React.ComponentType<IFormCreateProps & T>):
React.ComponentType<ISharedFormCreate & T> {
  return class extends ComponentBase<ISharedFormCreate & T, ISharedFormCreateState> {
    public submitCreate: SubmitCreate
    public constructor(props: ISharedFormCreate & T) {
      super(props)
      this.submitCreate = submitCreate(this)
    }
    public componentWillMount() {
      const prefix: any = get(this.props, "prefix", "")
      const valuePath =
        `${prefix && prefix.length > 0 ? prefix + "." : ""}${get(this.props, "form.model", "undefined")}`
      if (this.props.resetOnMount) {
        FormStore.resetForm(this.props.form, this.props.prefix)
        FormErrorStore.setError(valuePath, undefined)
      }
      if (this.props.resetState) {
        DbStore.setCreateState(valuePath, undefined)
      }
      super.componentWillMount()
    }
    public render() {
      return (
        <Component
          {...this.props}
          {...this.state}
          submitCreate={this.submitCreate}
          fields={filterFields(this.props.form, globals.Method.Create)}
        />
      )
    }
    protected _buildState(props: ISharedFormCreate, initialBuild: boolean): ISharedFormCreateState {
      const prefix = get(props, "prefix", "")
      const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${get(props, "form.model", "undefined")}`
      return {
        error: FormErrorStore.getError(valuePath),
        state: DbStore.getCreateState(valuePath),
        loading: LoadingStore.getLoading(valuePath),
        user: UserStore.getUser()
      }
    }
  }
}

export default SharedFormCreate
export {
  submitCreate
}
