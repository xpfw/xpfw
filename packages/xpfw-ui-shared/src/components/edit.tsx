import { FormErrorStore, FormStore, LoadingStore } from "@xpfw/form-shared"
import { filterFields, globals, IField, IForm, IFormError } from "@xpfw/validate"
import { get, isNil } from "lodash"
import * as React from "react"
import { ComponentBase } from "resub"
import DbStore from "../store/db"
import UserStore from "../store/user"

export interface ISharedFormEdit extends React.Props<any> {
  prefix?: string
  id?: string
  /**
   * Will reset the form values
   */
  resetOnMount?: boolean
  /**
   * Will reset the result of the last creation
   */
  resetState?: boolean
  form: IForm
}

export interface ISharedFormEditState {
  error?: IFormError
  state?: any
  original?: any
  user?: any
  loading: boolean
}

export type SubmitEdit = () => Promise<any>

export interface IFormEditProps extends ISharedFormEdit, ISharedFormEditState {
  submitEdit: SubmitEdit
  /** Prefiltered Fields to display */
  fields: IField[]
}

const submitEdit = (thisRef: {props: ISharedFormEdit}) => {
  return async () => {
    const prefix = get(thisRef, "props.prefix", "")
    await FormStore.validateForm(get(thisRef, "props.form"),
      globals.Method.Update, prefix, get(thisRef, "state.user"))
    const err =
    FormErrorStore.getError(`${prefix && prefix.length > 0 ? prefix + "." : ""}${get(thisRef, "props.form.model")}`)
    let res
    if (!isNil(err) && err.ok) {
      res = await DbStore.patch(get(thisRef, "props.id"),
        get(thisRef, "props.form"), get(thisRef, "props.prefix"))
    }
    return res
  }
}

function SharedFormEdit<T>(Component: React.ComponentType<IFormEditProps & T>):
React.ComponentType<ISharedFormEdit & T> {
  return class extends ComponentBase<ISharedFormEdit & T, ISharedFormEditState> {
    public submitEdit: SubmitEdit
    public constructor(props: ISharedFormEdit & T) {
      super(props)
      this.submitEdit = submitEdit(this)
    }
    public componentWillMount() {
      const prefix: any = get(this.props, "prefix", "")
      const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${get(this.props, "id", "")}${get(this.props, "form.model", "undefined")}`
      if (this.props.resetOnMount) {
        FormErrorStore.setError(valuePath, undefined)
      }
      if (this.props.resetState) {
        DbStore.setUpdateState(valuePath, undefined)
      }
      super.componentWillMount()
    }
    public render() {
      return (
        <Component
          {...this.props}
          {...this.state}
          submitEdit={this.submitEdit}
          fields={filterFields(this.props.form, globals.Method.Update)}
        />
      )
    }
    protected _buildState(props: ISharedFormEdit, initialBuild: boolean): ISharedFormEditState {
      const prefix: any = get(props, "prefix", "")
      const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${get(props, "id", "")}${get(props, "form.model", "undefined")}`
      const id: any = get(props, "id", "none")
      const collection = get(props, "form.collection", "none")
      return {
        error: FormErrorStore.getError(valuePath),
        state: DbStore.getUpdateState(valuePath),
        loading: LoadingStore.getLoading(valuePath),
        original: DbStore.getEditOriginal(id, get(props, "form"), get(props, "prefix", "")),
        user: UserStore.getUser()
      }
    }
  }
}

export default SharedFormEdit
export {
  submitEdit
}
