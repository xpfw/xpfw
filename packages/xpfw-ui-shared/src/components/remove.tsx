import { FormErrorStore, FormStore, LoadingStore } from "@xpfw/form-shared"
import { globals, IForm, IFormError } from "@xpfw/validate"
import { get } from "lodash"
import * as React from "react"
import { ComponentBase } from "resub"
import DbStore, { REMOVE_ADDON_KEY } from "../store/db"

export interface ISharedFormRemove extends React.Props<any> {
  id?: string
  form: IForm
}

export interface ISharedFormRemoveState {
  error?: IFormError
  state?: any
  loading: boolean
}

export type SubmitRemove = () => Promise<any>

export interface IFormRemoveProps extends ISharedFormRemove, ISharedFormRemoveState {
  submitRemove: SubmitRemove
}

const submitRemove = (thisRef: {props: ISharedFormRemove}) => {
  return async () => {
    return DbStore.remove(get(thisRef, "props.id"),
        get(thisRef, "props.form.collection"))
  }
}

function SharedFormRemove<T>(Component: React.ComponentType<IFormRemoveProps & T>):
React.ComponentType<ISharedFormRemove & T> {
  return class extends ComponentBase<ISharedFormRemove & T, ISharedFormRemoveState> {
    public submitRemove: SubmitRemove
    public constructor(props: ISharedFormRemove & T) {
      super(props)
      this.submitRemove = submitRemove(this)
    }
    public render() {
      return (
        <Component
          {...this.props}
          {...this.state}
          submitRemove={this.submitRemove}
        />
      )
    }
    protected _buildState(props: ISharedFormRemove, initialBuild: boolean): ISharedFormRemoveState {
      const id: any = get(props, "id", "_")
      return {
        state: DbStore.getRemoveState(id),
        loading: LoadingStore.getLoading(REMOVE_ADDON_KEY + id)
      }
    }
  }
}

export default SharedFormRemove
export {
  submitRemove
}
