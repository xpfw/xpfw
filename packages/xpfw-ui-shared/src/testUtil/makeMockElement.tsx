import { cloneDeep, get, isArray, isNil, isObject, isString, set } from "lodash"
import * as React from "react"

const idtmp: string[] = []

const makeMockElement = (name: string, stripIdsAt?: string | string[], idPath: string = "_id") => {
  return class extends React.Component<any, any> {
    public render() {
      const toStringify = this.props
      const stripMe = isArray(stripIdsAt) ? stripIdsAt : [stripIdsAt]
      if (isArray(stripIdsAt)) {
        for (const stripper of stripIdsAt) {
          if (isString(stripper)) {
            let toStrip = get(this.props, stripper)
            if (!isArray(toStrip)) {
              toStrip = [toStrip]
            }
            const newIds = []
            for (const obj of toStrip) {
              const id = isString(obj) ? obj : get(obj, idPath)
              let newId = idtmp.indexOf(id)
              if (newId === -1) {
                newId = idtmp.length + 1
                idtmp.push(id)
              }
              const objClone = cloneDeep(obj)
              if (isString(obj)) {
                newIds.push(newId)
              } else {
                set(objClone, idPath, newId)
                newIds.push(objClone)
              }
            }
            if (newIds.length > 0 && !isNil(newIds[0])) {
              set(toStringify, stripper, newIds)
            }
          }
        }
      }
      return (
        <div className={name}>
          {JSON.stringify(toStringify, undefined, 2)}
        </div>
      )
    }
  }
}

export default makeMockElement
