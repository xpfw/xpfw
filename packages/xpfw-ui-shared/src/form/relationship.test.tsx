import { ComponentRegistry, FormStore, SharedField } from "@xpfw/form-shared"
import { getRandomApp } from "@xpfw/test-util"
import { FeathersClient } from "@xpfw/ui-feathers"
import { FieldType, TestDefs } from "@xpfw/validate"
import "isomorphic-fetch"
import { get } from "lodash"
import * as render from "preact-render-to-json"
import * as React from "react"
import { matchStoreState } from "resub-persist"
import BackendClient from "../client"
import { submitCreate } from "../components/create"
import DbStore from "../store/db"
import makeMockElement from "../testUtil/makeMockElement"
import RelationShipWrapper, { addId, removeId, searchRelated } from "./relationship"

BackendClient.client = FeathersClient

test("Relationship Fetching Test", async () => {
  const form = TestDefs.FormNumberAndRequiredText
  const s: any = [TestDefs.FormNumberAndRequiredText.collection, TestDefs.FormRelationshipToNumAndRequiredText.collection]
  const relationCol: any = TestDefs.FormNumberAndRequiredText.collection
  const appRef = await getRandomApp(s, false, BackendClient.client, true)
  const MockEle: any = makeMockElement("show")
  const MockedRelationship = RelationShipWrapper(MockEle)
  ComponentRegistry.registerComponent(FieldType.RelationshipSingle, MockedRelationship)
  ComponentRegistry.registerComponent(FieldType.RelationshipMulti, MockedRelationship)
  const createdEles: any = []
  for (let i = 0; i < 10; i++) {
    FormStore.setValue(TestDefs.RequiredTextField.mapTo, "myText " + i)
    FormStore.setValue(TestDefs.NumberField.mapTo, 420 + i)
    createdEles.push(await submitCreate({props: {form}})())
  }

  // SINGLE

  expect(render(<SharedField field={TestDefs.RelationshipSingleField} />)).toMatchSnapshot("nothing set yet")
  FormStore.setValue(TestDefs.RelationshipSingleField.mapTo, "INVALID")
  expect(render(<SharedField field={TestDefs.RelationshipSingleField} />)).toMatchSnapshot("set to invalid value")
  FormStore.setValue(TestDefs.RelationshipSingleField.mapTo, createdEles[0].result.id)
  expect(render(<SharedField field={TestDefs.RelationshipSingleField} />)).toMatchSnapshot("set to first creation not loaded yet")
  await DbStore.getFromServer(createdEles[1].result.id, relationCol)
  FormStore.setValue(TestDefs.RelationshipSingleField.mapTo, createdEles[1].result.id)
  expect(render(<SharedField field={TestDefs.RelationshipSingleField} />)).toMatchSnapshot("set to first creation loaded")
  FormStore.setValue(TestDefs.RelationshipMultiField.mapTo, [createdEles[3].result.id, createdEles[4].result.id, createdEles[5].result.id])

  // Multi

  await DbStore.getFromServer(createdEles[3].result.id, relationCol)
  await DbStore.getFromServer(createdEles[5].result.id, relationCol)
  expect(render(<SharedField field={TestDefs.RelationshipMultiField} />)).toMatchSnapshot("multi load with one missing")
  await DbStore.getFromServer(createdEles[4].result.id, relationCol)
  expect(render(<SharedField field={TestDefs.RelationshipMultiField} />)).toMatchSnapshot("multi load with none missing")
  const thisRef: any = {props: {field: TestDefs.RelationshipMultiField}}
  const boundAddId = addId(thisRef)
  const boundRemoveId = removeId(thisRef)
  boundRemoveId(createdEles[5].result.id)
  expect(render(<SharedField field={TestDefs.RelationshipMultiField} />)).toMatchSnapshot("multi removed one")
  boundRemoveId(createdEles[3].result.id)
  expect(render(<SharedField field={TestDefs.RelationshipMultiField} />)).toMatchSnapshot("multi removed two")
  boundAddId(createdEles[7].result.id)
  expect(render(<SharedField field={TestDefs.RelationshipMultiField} />)).toMatchSnapshot("multi added but not yet loaded")
  await DbStore.getFromServer(createdEles[7].result.id, relationCol)
  expect(render(<SharedField field={TestDefs.RelationshipMultiField} />)).toMatchSnapshot("multi added and loaded")
  await appRef.cleanUp()
}, 10000)
