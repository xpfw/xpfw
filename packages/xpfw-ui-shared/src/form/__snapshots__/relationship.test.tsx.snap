// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Relationship Fetching Test: multi added and loaded 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedByMulti",
    "type": 14,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "value": [
    4,
    7
  ],
  "relatedObject": [
    {
      "myString": "myText 4",
      "myNum": 424,
      "id": 4
    },
    {
      "myString": "myText 7",
      "myNum": 427,
      "id": 7
    }
  ],
  "searchForm": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          }
        ]
      }
    ]
  },
  "displayMode": 0,
  "lastUsed": [
    7
  ]
}
</div>
`;

exports[`Relationship Fetching Test: multi added but not yet loaded 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedByMulti",
    "type": 14,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "value": [
    4,
    7
  ],
  "relatedObject": [
    {
      "myString": "myText 4",
      "myNum": 424,
      "id": 4
    },
    null
  ],
  "searchForm": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          }
        ]
      }
    ]
  },
  "displayMode": 0,
  "lastUsed": [
    7
  ]
}
</div>
`;

exports[`Relationship Fetching Test: multi load with none missing 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedByMulti",
    "type": 14,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "value": [
    3,
    4,
    5
  ],
  "relatedObject": [
    {
      "myString": "myText 3",
      "myNum": 423,
      "id": 3
    },
    {
      "myString": "myText 4",
      "myNum": 424,
      "id": 4
    },
    {
      "myString": "myText 5",
      "myNum": 425,
      "id": 5
    }
  ],
  "searchForm": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          }
        ]
      }
    ]
  }
}
</div>
`;

exports[`Relationship Fetching Test: multi load with one missing 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedByMulti",
    "type": 14,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "value": [
    3,
    4,
    5
  ],
  "relatedObject": [
    {
      "myString": "myText 3",
      "myNum": 423,
      "id": 3
    },
    null,
    {
      "myString": "myText 5",
      "myNum": 425,
      "id": 5
    }
  ],
  "searchForm": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          }
        ]
      }
    ]
  }
}
</div>
`;

exports[`Relationship Fetching Test: multi removed one 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedByMulti",
    "type": 14,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "value": [
    3,
    4
  ],
  "relatedObject": [
    {
      "myString": "myText 3",
      "myNum": 423,
      "id": 3
    },
    {
      "myString": "myText 4",
      "myNum": 424,
      "id": 4
    }
  ],
  "searchForm": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          }
        ]
      }
    ]
  }
}
</div>
`;

exports[`Relationship Fetching Test: multi removed two 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedByMulti",
    "type": 14,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "value": [
    4
  ],
  "relatedObject": [
    {
      "myString": "myText 4",
      "myNum": 424,
      "id": 4
    }
  ],
  "searchForm": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          }
        ]
      }
    ]
  }
}
</div>
`;

exports[`Relationship Fetching Test: nothing set yet 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedBy",
    "type": 13,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "searchForm": {
    "model": "searchFor.ownedBy",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          }
        ]
      }
    ]
  }
}
</div>
`;

exports[`Relationship Fetching Test: set to first creation loaded 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedBy",
    "type": 13,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "value": 1,
  "relatedObject": {
    "myString": "myText 1",
    "myNum": 421,
    "id": 1
  },
  "searchForm": {
    "model": "searchFor.ownedBy",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          }
        ]
      }
    ]
  }
}
</div>
`;

exports[`Relationship Fetching Test: set to first creation not loaded yet 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedBy",
    "type": 13,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "value": 0,
  "searchForm": {
    "model": "searchFor.ownedBy",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          }
        ]
      }
    ]
  }
}
</div>
`;

exports[`Relationship Fetching Test: set to invalid value 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedBy",
    "type": 13,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "value": "INVALID",
  "searchForm": {
    "model": "searchFor.ownedBy",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          }
        ]
      }
    ]
  }
}
</div>
`;
